beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+		s, err = ReplayProcessSlots(ctx, s, rs.blockSlot)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	duration := time.Since(start)
+	log.WithFields(logrus.Fields{
+		"duration": duration,
+	}).Debug("finished calling process_blocks on all blocks in ReplayBlocks")
+	return s, nil
+}
+
+// ReplayToSlot invokes ReplayBlocks under the hood,
+// but then also runs process_slots to advance the state past the root or slot used in the builder.
+// for example, if you wanted the state to be at the target slot, but only integrating blocks up to
+// slot-1, you could request Builder.ForSlot(slot-1).ReplayToSlot(slot)
+func (rs *stateReplayer) ReplayToSlot(ctx context.Context, target types.Slot) (state.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "stateGen.stateReplayer.ReplayToSlot")
+	defer span.End()
+
+	s, err := rs.ReplayBlocks(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to ReplayBlocks")
+	}
+
+	if target > s.Slot() {
+		start := time.Now()
+		log.WithFields(logrus.Fields{
+			"startSlot": s.Slot(),
+			"endSlot":   target,
+			"diff":      target - s.Slot(),
+		}).Debug("calling process_slots on remaining slots")
+
+		if rs.blockSlot > s.Slot() {
+			// err will be handled after the bookend log
+			s, err = ReplayProcessSlots(ctx, s, target)
+		}
+
+		duration := time.Since(start)
+		log.WithFields(logrus.Fields{
+			"duration": duration,
+		}).Debug("time spent in process_slots")
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("ReplayToSlot failed to seek to slot %d after applying blocks", target))
+		}
+	}
+	return s, nil
+}
+
+// ReplayerBuilder creates a Replayer that can be used to obtain a state at a specified slot or root
+// (only ForSlot implemented so far).
+// See documentation on Replayer for more on how to use this to obtain pre/post-block states
+type ReplayerBuilder interface {
+	// ForSlot creates a builder that will create a state that includes blocks up to and including the requested slot
+	// The resulting Replayer will always yield a state with .Slot=target; if there are skipped blocks
+	// between the highest canonical block in the db and the target, the replayer will fast-forward past the intervening
+	// slots via process_slots.
+	ForSlot(target types.Slot) Replayer
+}
+
+// NewCanonicalBuilder handles initializing the default concrete ReplayerBuilder implementation.
+func NewCanonicalBuilder(h HistoryAccessor, c CanonicalChecker, cs CurrentSlotter) *CanonicalBuilder {
+	return &CanonicalBuilder{
+		Chainer: &CanonicalChainer{
+			h:  h,
+			c:  c,
+			cs: cs,
+		},
+	}
+}
+
+type retrievalMethod int
+
+const (
+	forSlot retrievalMethod = iota
+	byRoot
+)
+
+// CanonicalBuilder builds a Replayer that uses a combination of database queries and a
+// CanonicalChecker (which should usually be a fork choice store implementing an IsCanonical method)
+// to determine the canonical chain and apply it to generate the desired state.
+type CanonicalBuilder struct {
+	Chainer Chainer
```

Comment:
Don't need to be exported?

```suggestion
	chainer Chainer
```

Reply:
agreed